### PR TITLE
Added faster AUROC

### PR DESCRIPTION
**Pull request by ShreyaR**
Attached some screenshots below of the comparison with `torchmetrics`' AUROC metric, and the maximum delta is around 0.005.

Screen Shot 2022-02-17 at 3 38 06 PM
Screen Shot 2022-02-17 at 3 39 35 PM
Screen Shot 2022-02-17 at 3 38 39 PM
Screen Shot 2022-02-17 at 3 38 29 PM

### CODE REVIEW

File: ludwig/modules/metric_modules.py [hunk 1]
```diff
@@ -123,11 +123,27 @@ def get_inputs(cls):
...
 
-    """ Metric for area under ROC curve. """
+    def __init__(
+        self,
+        num_thresholds: int = 201,
+        epsilon: float = 1e-7,
+        compute_on_step: Optional[bool] = None,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Optional[Callable] = None,
```
Comment (tgaddair) [inline ludwig/modules/metric_modules.py:136]: Should be `gather_all_tensors` (see `MeanMetric` just below).

File: ludwig/modules/metric_modules.py [hunk 2]
```diff
@@ -123,11 +123,27 @@ def get_inputs(cls):
...
-        super().__init__(dist_sync_fn=gather_all_tensors)
+class ROCAUCMetric(LudwigMetric):
+    """Fast implementation of metric for area under ROC curve."""
 
-    """ Metric for area under ROC curve. """
+    def __init__(
+        self,
+        num_thresholds: int = 201,
+        epsilon: float = 1e-7,
+        compute_on_step: Optional[bool] = None,
```
Comment (tgaddair) [inline ludwig/modules/metric_modules.py:133]: The generic torchmetrics params should be capable of being passed through the `**kwargs`, so shouldn't need to specify explicitly.

**Review by tgaddair — COMMENTED** [review]
Amazing! Just a couple points regarding the `__init__` params.

File: ludwig/modules/metric_modules.py (hunk 1 above)
Comment (ShreyaR) [reply ludwig/modules/metric_modules.py]: Done!

File: ludwig/modules/metric_modules.py (hunk 2 above)
Comment (ShreyaR) [reply ludwig/modules/metric_modules.py]: Updated the code to follow a similar pattern as what the other metrics are doing. It looks like we pass in some common arguments to each metric module in **kwargs, so I'm not passing in all kwargs to the torchmetrics constructor.

File: ludwig/modules/metric_modules.py [hunk 3]
```diff
@@ -123,11 +123,26 @@ def get_inputs(cls):
...
+    def __init__(
+        self,
+        num_thresholds: int = 201,
+        epsilon: float = 1e-7,
+        **kwargs,
+    ) -> None:
         super().__init__(dist_sync_fn=gather_all_tensors)
+        self.num_thresholds = num_thresholds
+        self.epsilon = epsilon
+        # Add an extra dimension to each state to make it compatible with Horovod all gather.
```
Comment (tgaddair) [inline ludwig/modules/metric_modules.py:138]: What do you think about moving this fix to `gather_all_tensors`? It seems the correct behavior for all metrics should be something like:

```
t = t.reshape([1, *t.shape])
hvd.allgather(t)
```

Comment (ShreyaR) [reply ludwig/modules/metric_modules.py]: That makes sense -- Ip updated the code to reflect this.

I guess we weren't seeing issues from this previously as each metric being gathered was a scalar value, and so concatenating and reducing them wasn't causing an issue.

I haven't tested out this change yet but I can update here to make sure that it doesn't cause an issue with other metrics.

**Review by tgaddair — APPROVED** [review]
LGTM!

**Pull request merged**